pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +241,289 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+					commitRec = &kgo.Record{
+						Topic:       state.Commit.Topic,
+						Partition:   state.Commit.Partition,
+						Offset:      state.Commit.At - 1, // Previous commit's offset-1 means we expect the next cycle to start from the commit's offset.
+						Timestamp:   state.CommitRecTimestamp,
+						LeaderEpoch: state.Commit.LeaderEpoch,
+					}
+				} else if commitRec == nil {
+					// The commit offset should be the last record that was fully processed and not the first record that was not fully processed.
+					commitRec = lastRec
+				}
+			}
+			lastRec = rec
+		}
+	}
+
+	// Nothing was consumed from Kafka at all.
+	if firstRec == nil {
+		level.Info(partitionLogger).Log("msg", "no records were consumed")
+		return nil
+	}
+
+	// No records were processed for this cycle.
+	if lastRec == nil {
+		level.Info(partitionLogger).Log("msg", "nothing to commit due to first record fetched is from next cycle", "first_rec_offset", firstRec.Offset)
+		return nil
+	}
+
+	// All samples in all records were processed. We can commit the last record's offset.
+	if commitRec == nil {
+		commitRec = lastRec
+	}
+
+	var err error
+	numBlocks, err = builder.CompactAndUpload(ctx, b.uploadBlocks)
+	if err != nil {
+		return err
+	}
+
+	commitRecOffset := commitRec.Offset + 1 // offset+1 means everything up to (including) the offset was processed
+
+	// We should take the max of "seen till" offset. If the partition was lagging
+	// due to some record not being processed because of a future sample, we might be
+	// coming back to the same consume cycle again.
+	oldestSeenOffset := max(lastRec.Offset, state.OldestSeenOffset)
+	// Take the max of block max times because of the same reasons above.
+	lastBlockEnd := blockEnd
+	if lastBlockEnd.Before(state.LastBlockEnd) {
+		lastBlockEnd = state.LastBlockEnd
+	}
+
+	commit := kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRecOffset,
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), oldestSeenOffset, lastBlockEnd.UnixMilli()),
+	}
+	if err := b.commitOffset(ctx, partitionLogger, b.cfg.Kafka.ConsumerGroup, commit); err != nil {
+		return err
+	}
+
+	// The new lag is the distance between fully processed offsets. If the partition's cycle was chopped into sections due to lagging,
+	// the cycle will continue consuming the next portion of the lag.
+	state.Lag -= commitRecOffset - firstRec.Offset
+
+	state.OldestSeenOffset = oldestSeenOffset
+	state.LastBlockEnd = lastBlockEnd
+	state.Commit = commit
+
+	return nil
+}
+
+func (b *BlockBuilder) seekPartition(partition int32, offset int64) {
+	offsets := map[string]map[int32]kgo.EpochOffset{
+		b.cfg.Kafka.Topic: {
+			partition: {
+				Epoch:  -1, // Epoch doesn't play any role outside the group consumption.
+				Offset: offset,
+			},
+		},
+	}
+	b.kafkaClient.SetOffsets(offsets)
```

Comment:
the godoc on SetOffsets says that you must've used the same client to consume from that partition to be able to adequately set the offset. From what I could tell you're not guaranteed this, right? 

I believe that's the reason we used two clients in the ingest storage ingester

Reply:
> the godoc on SetOffsets says that you must've used the same client to consume from that partition to be able to adequately set the offset.

I think you refer to [this part][1] 👇 right?

> it is invalid to set topics that were not yet returned from a PollFetches: this function sets only partitions that were previously consumed, any extra partitions are skipped.

I don't think this caveat is meaningful to us. Here we only need to move the client's internal cursor to the offset of the last commit, in case the client overconsumed on the previous cycle-section.

On the very first call to `consumePartitionCycle` after the rest, this is a noop, because the client didn't consume anything, so it will consume from the commit (or, if there is no commit, it will reset to the fallback, due to `kgo.ConsumeResetOffset` option).

Effectively, what we are doing here is as if we re-created a brand-new Kafka client, every time we enter `consumePartitionCycle`.

[1]: https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#Client.SetOffsets

Reply:
> It will start consuming from last commit

what guarantees this? Since we're consuming out of a consumer group (i.e. not relying on the kafka client to manage the commits), we manage the start offsets. Then doesn't that mean that we'll always offset to `fallbackOffsetMillis` the first time we start consuming a partition?

Reply:
> doesn't that mean that we'll always offset to `fallbackOffsetMillis` the first time we start consuming a partition?

That's a good point to check. In our unit-tests, I saw that it behaved as expected (i.e. the [`TestBlockBuilder_StartWithExistingCommit`](https://github.com/grafana/mimir/blob/cd35789928f6d4763526cea84ec235046197be43/pkg/blockbuilder/blockbuilder_test.go#L95) test consumes from the last commit). But since the past refactoring in the backporting PRs, I also haven't tested it against the real Kafka with existing data. I should do that.

Reply:
I've reworked this part, see https://github.com/grafana/mimir/pull/9199/commits/d3786fdf03894049e43505fe4ddafd3d37ac2630.

This was indeed a bug that we introduced during the refactorings between #9068 and #9147. `kfake` didn't show it, because it behaves differently in when partitions and offsets are assigned (I haven't delved into the roots of it). The new version works in real Kafka, though.